### Problem 1

**Note:** This assignment is due on Sunday, April 25th at Noon ET and will be graded Pass/Fail.  No extensions allowed.

Dear Students,

Welcome to the last homework in DATA2040-sp21.  You can work with up to 1 additional partner on this assignment; only one partner may submit if you choose to do this. 
 

The exercises in this homework are designed to help ensure you understand the setup for pre-training a BERT model and how to fine-tune BERT models for specific applications.
 
After completing this assignment, you’ll understand the connection between BERT models and BERT applications and have the know-how to perform fine-tuning for Extractive Question and Answering (EQA) using a Keras head model (running on GPUs and TPUs) and with the standard HuggingFace.co transformer library (running on GPUs).
 
Along the way, you’ll demonstrate your understanding of the data preprocessing methods required for both BERT and EQA and your knowledge of applicable training loss functions and metrics.
 
**Primary Resource**
This paper [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805) introduced BERT to the world. Please use it as your primary reference while working on this assignment, i.e., try and work through the relevant parts. 

Why? Because papers like this one are the format in which most new Deep Learning techniques are introduced. It often takes several years (and sometimes decades) before secondary materials become available for the methods that become recognized as important.

**Additional Resources **
Once you've taken a shot at it, please feel free to access other materials and google.  Below is a curated list of suggestions.

First up is the paper's official github [repo](https://github.com/google-research/bert).  Mapping papers to code, and code to papers is a great way to improve your practical DL skills. The code is well organized, but complex; here is one file that may be particularly helpful to took at  [run_pretraining.py](https://github.com/google-research/bert/blob/eedf5716ce1268e56f0a50264a88cafad334ac61/run_pretraining.py#L131)

Next up areJay Alammar's blog posts on Transformers and BERT:

- [The Illustrated Transformer](http://jalammar.github.io/illustrated-transformer/)
- [The Illustrated BERT, ELMo, and co. (How NLP Cracked Transfer Learning)](http://jalammar.github.io/illustrated-bert/)


**The Future of DL (and Data Science)**
Up to now, understanding what's been learned by complex NLP models has been a challenge.  Alammar's Pyhton [Ecco](https://www.eccox.io/) package is one attempt to change that. It provides visualizations of the information in Transformer models. These two blog posts provide an introduction to the work so far:

- [Interfaces for Explaining Transformer Language Models](http://jalammar.github.io/explaining-transformers/)
- [Find the Words to Say: Hidden State Visualization for Language Models](http://jalammar.github.io/hidden-states/)


People that have a deep understanding of complex structures are in a good position to create innovative applications based on them and to perhaps develop new and important structures based on this knowledge.  The development of BERT is a perfect example.  


Good luck wrapping up your Final Projects and this assignment.
 
On behalf of the entire course staff, I would like to thank you for participating in DATA 2040. It has been a pleasure to have you all as students.

Best Wishes,
Dan Potter


**Partner Option**
You may work with up to 1 additional partner on this assignment; only one partner may submit if you choose to do this. 
**®Task: **If you worked with a partner on this assignment, please list their official email address below.


### Problem 2

**Partner Option**
**®Task:** If you worked with a partner, please confirm that you and your partner both understand that only one of you may submit this homework.


### Problem 3

**Pretraining BERT**

**®[15] Task: A is for Attention, B is for Bert, C is for Cross-Entropy Loss **
Describe how BERT is pretrained.  
1. Include an explanation of the what training inputs look like, including the labels that are learned.
2. As part of this, include a few specific training input examples.  Include details on any special tokens used (like [CLS], [SEP], [MASK]), and what the corresponding labels to learn are.


In general, the BERT is trained using two unsupervised task, Masked Language Modeling and Next Sentence Prediction. The loss used here is Cross Entropy to between the labels and predictions. 

1. To make BERT handle a variety of down-stream tasks, our input representation is able to unambiguously represent both a single sentence and a pair of sentences in one token sequence. Throughout this work, a “sentence” can be an arbitrary span of contiguous text, rather than an actual linguistic sentence. A “sequence” refers to the input token sequence to BERT, which may be a single sentence or two sentences packed together. The input embeddings are the sum of the token embeddings, the segmentation embeddings and the position embeddings.
The training input data includes BooksCorpus and English WikiPedia text messages. The start of the sentence is with [CLS], masked tokens with [MASK] and seperate between sequences with [SEP]. Also, the padding is used to have fixed length.

2. 15% of the word tokens in the sequence will be masked and 80% of the time, the token is masked with special token [MASK], 10% of the time unchanged and 10% of the time with random token to force BERT to learn from the previous and later tokens instead of learning from the word itself. The training use Cross-Entropy Loss.
Also for NSP, 50% of the time, the sentences are in order and 50% percent of the time the second sentence is actually not the next one.

Masked LM: \\
The input: [CLS] the man went to [MASK] store [SEP] he bought a bottle of [MASK] coke [SEP] \\
The output: the diet \\
The input: [CLS] I love reading [MASK] \\
The output: books 

NSP:\\
The input: [CLS] I am a Brown [MASK] [SEP] I study [MASK] science [SEP] \\
The output: IsNext \\
The input: [CLS] The day is [MASK] [SEP] NBA pauses due to [Mask] [SEP] \\
The output: NotNext


### Problem 4

**Pretraining BERT**

**®[5] Task: Attention to detailZzzz**
How is the loss function defined?  Since BERT is trained on more than one task, how are they combined?


The training loss is the sum of the mean masked LM likelihood and the mean next sentence prediction likelihood.

### Problem 5

**Using pre-trained BERT models**

**®[10] Task: Machine Intelligence?**
The BERT paper explains how to configure and finetune a pre-trained BERT model to perform various NLP tasks. 
1. What are the specific tasks described? 
2. At which tasks was BERT documented as being state of the art?


1. GLUE (General Language Understanding Evaluation), SQuAD v1.1 (Stanford Question Answering Dataset),  SQuAD v2.0 and SWAG (Situations With Adversarial Generations).

2. SQuAD v2.0, MNLI (Multi-Genre Natural Language Inference), QQP (Quora Question Pairs), QNLI (Question Natural Language Inference ), SST-2 (The Stanford Sentiment Treebank), CoLA (The Corpus of Linguistic Acceptability), STS-B (The Semantic Textual Similarity Benchmark), MRPC (Microsoft Research Paraphrase Corpus), RTE (Recognizing Textual Entailment), WNLI (Winograd NLI), SWAG (Situations With Adversarial Generations). In total, BERT advanced 11 NLP tasks.

### Problem 6

**Using pre-trained BERT models**

**®[5] Task: School versus learning a new task**
Which takes longer, pre-training or fine-tuning? Why?


Pre-training. Since fine-tuning of BERT is sensitive to small datasets than large-scaled datasets, and only a few parameters will be fine-tuned during the process. Also, hyperparameters are chosen in a given range with most of them same as those in pre-training. Furthermore, fine-tuning only cares about a specific task and uses features generated from pre-training to solve speicfic problems. Therefore, fine-tuning is faster than pre-training.

### Problem 7

**Extractive question and answering**
Please review the following extractive question and answer (EQA) problems before beginning to answer them.  Each one will be graded on the information requested. 

**®[15] Task: Focus on the answer**
1. What is extractive question and answering? 
2. What is [SQuAD1.1](https://rajpurkar.github.io/SQuAD-explorer/explore/1.1/dev/)? 
3. Provide an example of the task.


1. Given a question and passage, the model should predict the answer or provide the answer text to that question.

2. SQuAD1.1 is a collection of 100k crowdsourced question/answer pairs. The answer in the label is directly cutted from a part of the passages. Three ground truth answers will usually contain three different labels (same if no enough).

3. Question: What is a course of study called? \\
Passage: The role of teacher is often formal and ongoing, carried out at a school or other place of formal education. In many countries, a person who wishes to become a teacher must first obtain specified professional qualifications or credentials from a university or college. These professional qualifications may include the study of pedagogy, the science of teaching. Teachers, like other professionals, may have to continue their education after they qualify, a process known as continuing professional development. Teachers may use a lesson plan to facilitate student learning, providing a course of study which is called the curriculum.\\
Ground Truth Answers: the curriculum. curriculum curriculum

### Problem 8

**Extractive question and answering**

**®Task [10]: Teaching BERT to do EQA #1**
Describe the BERT fine-tuning process for extractive question and answering.  Assume the SQuAD1.1 dataset is being used.


The question and paragraph are combined in to a single pack as input into BERT with different embeddings for their identities. S and E for start and end are introduced for the output to limit a span of text from passage. The probability of word i being the
start of the answer span is computed as a dot product between Ti and S followed by a softmax over all of the words in the paragraph. The training objective is the sum of the log-likelihoods of the correct start and end positions. The developers fine-tune for 3 epochs with a learning rate of 5e-5 and a batch size of 32. They also use data augmentation by fine-tuning TriviaQA before fine-tuning with SQuAD.

### Problem 9

**Extractive question and answering**
**®[10] Task: Teaching BERT to do EQA #2**
Provide two training input examples, and details on [CLS], [SEP], [MASK], and labels. It would be enough if the format is correct. 
(Hint: you can find relevant information in the [BERT](https://arxiv.org/pdf/1810.04805.pdf) paper)


1. [CLS] who is dog [sep] alice needs to walk dog tom every afternoon [sep]

2. [CLS] who studies at brown [sep] alice visits tom at brown in the last summer [sep]

### Problem 10

**Extractive question and answering**
**®[10] Task: Teaching BERT to do EQA #3**
Confirm your training sample input understanding by running them through the the preprocessing code in this [notebook](https://keras.io/examples/nlp/text_extraction_with_bert/) (keras.io). 

Copy and paste the output of your computer experiment below and explain it. (Hint: you may find that the paper and the code change the order of some things, that is OK)


This is the code I used to go through the process and check output:\\

```
squad_eg = SquadExample('who is dog', 'alice needs to walk dog tom every afternoon', 24, ['tom'], [['tom']])
squad_eg.preprocess()
tokenizer.decode(squad_eg.input_ids)
squad_eg.input_ids
```
The decoded output: alice needs to walk dog tom every afternoon who is dog\\
The encoded output: [101,5650,3791,2000,3328,3899,3419,2296,5027,102,2040,2003,
3899,102,0,0,0,0,0...] \\
We can see that the 101 means start [CLS] and 102 is the token for [SEP]. The question and context are merged into one input.


```
squad_eg = SquadExample('who studies at brown', 'alice visits tom at brown in the last summer', 13, ['tom'], [['tom']])
squad_eg.preprocess()
squad_eg.input_ids
```
The decoded output: alice visits tom at brown in the last summerwho studies at brown \\
The encoded output: [101,5650,7879,3419,2012,2829,1999,1996,2197,2621,102,2040,2913,2012,2829,102,0,0,0,0...] \\
We can see that the 101 means start [CLS] and 102 is the token for [SEP]. The question and context are merged into one input.




### Problem 11

**Extractive question and answering**
**®[5] Task: Teaching BERT to do EQA #4**
The pre-trained BERT model has multiple outputs; which ones are used for fine-tuning the squad task and which ones are ignored?
(Hint: you can find relevant information in the [BERT](https://arxiv.org/pdf/1810.04805.pdf) paper)


The probability of word i being the start of the answer span is computed as a dot product between Ti and S followed by a softmax over all of the words in the paragraph. Take the state of last hidden layer and feed it into the start token classifier. The start token classifier only has a single set of weights which it applies to every word. After taking the dot product between the output embeddings and the start weights, it applies the softmax activation to produce a probability distribution over all of the words. Whichever word has the highest probability of being the start token is the one that we pick. It also repeats this process for the end token — we have a separate weight vector for this.

### Problem 12

**Extractive question and answering**

**®[5] Task: Teaching BERT to do EQA #5**
Identify and describe the loss function that is used. Provide sufficient detail for the reader to understand why it’s appropriate.


Code for loss in github:
```
 def compute_loss(logits, positions):
        one_hot_positions = tf.one_hot(
            positions, depth=seq_length, dtype=tf.float32)
        log_probs = tf.nn.log_softmax(logits, axis=-1)
        loss = -tf.reduce_mean(
            tf.reduce_sum(one_hot_positions * log_probs, axis=-1))
        return loss
```


The training objective is the sum of the log-likelihoods of the correct start and end positions. Since the answer in SQuAD is a text span that comes from the paragraph, we only need to specify the start and end character index in the paragraph to describe the label and output. Under this schema, the loss is the average of the start_position loss start_loss and the end_postition loss end_loss. Each loss is computed in the same way: after applying the Softmax to the final output (i.e S*Ti or E*Ti) we use the real start/end postitions to compute the loss (see code below).

### Problem 13

**Extractive question and answering**
The keras.io notebook includes code to take advantage of TPU’s. It requires the model weights to be represented explicitly in the tensorflow model graph. 

**®[5] Task: Run with the wind #1**
Copy and paste the code cell(s) that expose the head model and compile it within a TPU context.


In [1]:
!pip install tokenizers
!pip install transformers

     |████████████████████████████████| 3.3MB 17.4MB/s 
     |████████████████████████████████| 2.1MB 16.7MB/s 
     |████████████████████████████████| 870kB 40.0MB/s 
  Created wheel for sacremoses: filename=sacremoses-0.0.44-cp37-none-any.whl size=886084 sha256=8f810937d7f31b7b1cb56ab86cc77e4a5fab922a281a1a18fa3f480f0ac03292
  Stored in directory: /root/.cache/pip/wheels/3e/fb/c0/13ab4d63d537658f448366744654323077c4d90069b6512f3c
Successfully built sacremoses


In [2]:
import os
import re
import json
import string
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tokenizers import BertWordPieceTokenizer
from transformers import BertTokenizer, TFBertModel, BertConfig
max_len = 384
configuration = BertConfig()

def create_model():
    ## BERT encoder
    encoder = TFBertModel.from_pretrained("bert-base-uncased")

    ## QA Model
    input_ids = layers.Input(shape=(max_len,), dtype=tf.int32)
    token_type_ids = layers.Input(shape=(max_len,), dtype=tf.int32)
    attention_mask = layers.Input(shape=(max_len,), dtype=tf.int32)
    embedding = encoder(
        input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
    )[0]

    start_logits = layers.Dense(1, name="start_logit", use_bias=False)(embedding)
    start_logits = layers.Flatten()(start_logits)

    end_logits = layers.Dense(1, name="end_logit", use_bias=False)(embedding)
    end_logits = layers.Flatten()(end_logits)

    start_probs = layers.Activation(keras.activations.softmax)(start_logits)
    end_probs = layers.Activation(keras.activations.softmax)(end_logits)

    model = keras.Model(
        inputs=[input_ids, token_type_ids, attention_mask],
        outputs=[start_probs, end_probs],
    )
    loss = keras.losses.SparseCategoricalCrossentropy(from_logits=False)
    optimizer = keras.optimizers.Adam(lr=5e-5)
    model.compile(optimizer=optimizer, loss=[loss, loss])
    return model

use_tpu = False # change this to use tpu or gpu
if use_tpu:
    # Create distribution strategy
    tpu = tf.distribute.cluster_resolver.TPUClusterResolver()
    tf.config.experimental_connect_to_cluster(tpu)
    tf.tpu.experimental.initialize_tpu_system(tpu)
    strategy = tf.distribute.experimental.TPUStrategy(tpu)

    # Create model
    with strategy.scope():
        model = create_model()
else:
    model = create_model()

model.summary()

Some layers from the model checkpoint at bert-base-uncased were not used when initializing TFBertModel: ['mlm___cls', 'nsp___cls']
- This IS expected if you are initializing TFBertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
All the layers of TFBertModel were initialized from the model checkpoint at bert-base-uncased.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertModel for predictions without further training.


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: module, class, method, function, traceback, frame, or code object was expected, got cython_function_or_method
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: module, class, method, function, traceback, frame, or code object was expected, got cython_function_or_method
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Cause: while/else statement not yet supported
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Cause: while/else statement not yet supported
To silence this warning, decorate the function with @tf.autograph.experi

### Problem 14

**Extractive question and answering**

**®[5] Task: Run with the wind #2 **
Using a batch size of 16, try fine-tuning with TPU vs. GPU enabled. Report the speed change below. How much GPU memory is used with a batch size of 16?  

Hints:
1. If you setup wandb.ai before you begin your experiments, you can use the System panel  ![](https://firebasestorage.googleapis.com/v0/b/prismia.appspot.com/o/user-images%252Fimage-76da31d9-93b6-49ae-a0ca-356e26e85e8f.png?alt=media&token=205261cb-6984-4246-b866-938082d192c1) to display system resource usage over time.
2. GPUs are slower than TPUs for this task.


In [3]:
slow_tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
save_path = "bert_base_uncased/"
if not os.path.exists(save_path):
    os.makedirs(save_path)
slow_tokenizer.save_pretrained(save_path)

# Load the fast tokenizer from saved file
tokenizer = BertWordPieceTokenizer("bert_base_uncased/vocab.txt", lowercase=True)

train_data_url = "https://rajpurkar.github.io/SQuAD-explorer/dataset/train-v1.1.json"
train_path = keras.utils.get_file("train.json", train_data_url)
eval_data_url = "https://rajpurkar.github.io/SQuAD-explorer/dataset/dev-v1.1.json"
eval_path = keras.utils.get_file("eval.json", eval_data_url)

class SquadExample:
    def __init__(self, question, context, start_char_idx, answer_text, all_answers):
        self.question = question
        self.context = context
        self.start_char_idx = start_char_idx
        self.answer_text = answer_text
        self.all_answers = all_answers
        self.skip = False

    def preprocess(self):
        context = self.context
        question = self.question
        answer_text = self.answer_text
        start_char_idx = self.start_char_idx

        # Clean context, answer and question
        context = " ".join(str(context).split())
        question = " ".join(str(question).split())
        answer = " ".join(str(answer_text).split())

        # Find end character index of answer in context
        end_char_idx = start_char_idx + len(answer)
        if end_char_idx >= len(context):
            self.skip = True
            return

        # Mark the character indexes in context that are in answer
        is_char_in_ans = [0] * len(context)
        for idx in range(start_char_idx, end_char_idx):
            is_char_in_ans[idx] = 1

        # Tokenize context
        tokenized_context = tokenizer.encode(context)

        # Find tokens that were created from answer characters
        ans_token_idx = []
        for idx, (start, end) in enumerate(tokenized_context.offsets):
            if sum(is_char_in_ans[start:end]) > 0:
                ans_token_idx.append(idx)

        if len(ans_token_idx) == 0:
            self.skip = True
            return

        # Find start and end token index for tokens from answer
        start_token_idx = ans_token_idx[0]
        end_token_idx = ans_token_idx[-1]

        # Tokenize question
        tokenized_question = tokenizer.encode(question)

        # Create inputs
        input_ids = tokenized_context.ids + tokenized_question.ids[1:]
        token_type_ids = [0] * len(tokenized_context.ids) + [1] * len(
            tokenized_question.ids[1:]
        )
        attention_mask = [1] * len(input_ids)

        # Pad and create attention masks.
        # Skip if truncation is needed
        padding_length = max_len - len(input_ids)
        if padding_length > 0:  # pad
            input_ids = input_ids + ([0] * padding_length)
            attention_mask = attention_mask + ([0] * padding_length)
            token_type_ids = token_type_ids + ([0] * padding_length)
        elif padding_length < 0:  # skip
            self.skip = True
            return

        self.input_ids = input_ids
        self.token_type_ids = token_type_ids
        self.attention_mask = attention_mask
        self.start_token_idx = start_token_idx
        self.end_token_idx = end_token_idx
        self.context_token_to_char = tokenized_context.offsets


with open(train_path) as f:
    raw_train_data = json.load(f)

with open(eval_path) as f:
    raw_eval_data = json.load(f)


def create_squad_examples(raw_data):
    squad_examples = []
    for item in raw_data["data"]:
        for para in item["paragraphs"]:
            context = para["context"]
            for qa in para["qas"]:
                question = qa["question"]
                answer_text = qa["answers"][0]["text"]
                all_answers = [_["text"] for _ in qa["answers"]]
                start_char_idx = qa["answers"][0]["answer_start"]
                squad_eg = SquadExample(
                    question, context, start_char_idx, answer_text, all_answers
                )
                squad_eg.preprocess()
                squad_examples.append(squad_eg)
    return squad_examples


def create_inputs_targets(squad_examples):
    dataset_dict = {
        "input_ids": [],
        "token_type_ids": [],
        "attention_mask": [],
        "start_token_idx": [],
        "end_token_idx": [],
    }
    for item in squad_examples:
        if item.skip == False:
            for key in dataset_dict:
                dataset_dict[key].append(getattr(item, key))
    for key in dataset_dict:
        dataset_dict[key] = np.array(dataset_dict[key])

    x = [
        dataset_dict["input_ids"],
        dataset_dict["token_type_ids"],
        dataset_dict["attention_mask"],
    ]
    y = [dataset_dict["start_token_idx"], dataset_dict["end_token_idx"]]
    return x, y


train_squad_examples = create_squad_examples(raw_train_data)
x_train, y_train = create_inputs_targets(train_squad_examples)
print(f"{len(train_squad_examples)} training points created.")

eval_squad_examples = create_squad_examples(raw_eval_data)
x_eval, y_eval = create_inputs_targets(eval_squad_examples)
print(f"{len(eval_squad_examples)} evaluation points created.")


4857856/4854279 [==============================] - 0s 0us/step
87599 training points created.
10570 evaluation points created.


In [4]:
def normalize_text(text):
    text = text.lower()

    # Remove punctuations
    exclude = set(string.punctuation)
    text = "".join(ch for ch in text if ch not in exclude)

    # Remove articles
    regex = re.compile(r"\b(a|an|the)\b", re.UNICODE)
    text = re.sub(regex, " ", text)

    # Remove extra white space
    text = " ".join(text.split())
    return text


class ExactMatch(keras.callbacks.Callback):
    """
    Each `SquadExample` object contains the character level offsets for each token
    in its input paragraph. We use them to get back the span of text corresponding
    to the tokens between our predicted start and end tokens.
    All the ground-truth answers are also present in each `SquadExample` object.
    We calculate the percentage of data points where the span of text obtained
    from model predictions matches one of the ground-truth answers.
    """

    def __init__(self, x_eval, y_eval):
        self.x_eval = x_eval
        self.y_eval = y_eval

    def on_epoch_end(self, epoch, logs=None):
        pred_start, pred_end = self.model.predict(self.x_eval)
        count = 0
        eval_examples_no_skip = [_ for _ in eval_squad_examples if _.skip == False]
        for idx, (start, end) in enumerate(zip(pred_start, pred_end)):
            squad_eg = eval_examples_no_skip[idx]
            offsets = squad_eg.context_token_to_char
            start = np.argmax(start)
            end = np.argmax(end)
            if start >= len(offsets):
                continue
            pred_char_start = offsets[start][0]
            if end < len(offsets):
                pred_char_end = offsets[end][1]
                pred_ans = squad_eg.context[pred_char_start:pred_char_end]
            else:
                pred_ans = squad_eg.context[pred_char_start:]

            normalized_pred_ans = normalize_text(pred_ans)
            normalized_true_ans = [normalize_text(_) for _ in squad_eg.all_answers]
            if normalized_pred_ans in normalized_true_ans:
                count += 1
        acc = count / len(self.y_eval[0])
        print(f"\nepoch={epoch+1}, exact match score={acc:.2f}")

In [5]:
import time
exact_match_callback = ExactMatch(x_eval, y_eval)
start = time.time()
model.fit(
    x_train,
    y_train,
    epochs=1,  # For demonstration, 3 epochs are recommended
    verbose=2,
    batch_size=16,
    callbacks=[exact_match_callback],
)
end = time.time()
print(end - start)

KeyboardInterrupt: ignored

In [6]:
end = time.time()
print(end - start)

6093.306775808334


Time with tpu: 518.122209072113 seconds \\
Time with gpu: I did not finish running within 1 hour. So I just print out the end time right after I interrupt the running and calculate the time elapsed: 6093.306775808334 seconds \\
It uses 6.06 GB memory and 14 GB GPU disk.

### Problem 15

**HuggingFace in practice**

Hugging face[ pipelines](https://huggingface.co/transformers/task_summary.html#extractive-question-answering) are an effortless way to use pre-trained and pre-finetuned models. As we saw in the last homework, Hugging Face also provides tools for fine-tuning.

**®[5] Task:  It's time to Hang 10**
Run the** **question and answer fine-tuning notebook from [The Big Table of Tasks](https://huggingface.co/transformers/examples.html?highlight=big%20table%20tasks#the-big-table-of-tasks) and run it. In the cell below, provide the final exact match and f-1 scores and explain what they mean.


{'exact_match': 76.74550614947965, 'f1': 85.13412652023338}

These scores are computed on individual question+answer pairs. When multiple correct answers are possible for a given question, the maximum score over all possible correct answers is computed. Overall EM and F1 scores are computed for a model by averaging over the individual example scores.

**Exact Match:** This metric is as simple as it sounds. For each question+answer pair, if the characters of the model's prediction exactly match the characters of (one of) the True Answer(s), EM = 1, otherwise EM = 0.

**F1 Score:** In this case, it's computed over the individual words in the prediction against those in the True Answer. The number of shared words between the prediction and the truth is the basis of the F1 score: precision is the ratio of the number of shared words to the total number of words in the prediction, and recall is the ratio of the number of shared words to the total number of words in the ground truth.

### Problem 16